<div dir="rtl" style="text-align:right">
<h1 id="دفتر-حصة-التمرين-1">دفتر حصة التمرين — 1</h1>
<p dir="rtl" style="text-align:right">نقرأ الشرح، ونتوقّع النتيجة، ثم ننفّذ الخلية التالية ونفسّر مخرجاتها قبل أن نتابع. نستخدم <strong>Restart Kernel and Run All</strong> لنتأكد من أن العمل لا يعتمد على جلسة سابقة. تبقى الشروح والاشتقاقات الرياضية والافتراضات الفيزيائية من الفصل محيطة بالخلايا.</p>
</div>

<div dir="rtl" style="text-align:right">
<h2 id="تهيئة-الدفتر-المرفقة">تهيئة الدفتر المرفقة</h2>
<p dir="rtl" style="text-align:right">نختار نواة <strong>Julia 1.12</strong> ونفتح الدفتر من مجلد فصله. تعثر هذه التهيئة المرفقة على بيئة المقرر المسجّلة وتنشئ مجلدًا منفصلًا للأشكال المصدّرة؛ وهي لا تثبّت أي حزم. أوامر التهيئة بنية تحتية، وليست مادة لغوية جديدة تخضع للتقييم.</p>
</div>

In [1]:
import Pkg
course_root = abspath("../..")
if !isfile(joinpath(course_root, "Project.toml"))
    error("Start this notebook from its week folder, e.g. week-1/notebooks")
end
Pkg.activate(course_root; io=devnull)
output_dir = get(ENV, "COURSE_NOTEBOOK_OUTPUT", joinpath(pwd(), "generated", "recitation-ar"))
mkpath(output_dir)
println("Course environment active; figure output directory ready.")

Course environment active; figure output directory ready.


<div dir="rtl" style="text-align:right">
<h1 id="تعابير-julia-الأولى-وملاءمة-نابض">تعابير <code dir="ltr">Julia</code> الأولى وملاءمة نابض</h1>
<p dir="rtl" style="text-align:right"><em>استوديو الأسبوع 1: نقرأ جدول نابض مُعطى، ونلائم ميله، ثم نبدأ الواجب الأساسي</em></p>
<h2 id="مرجع-سريع-من-شاشة-واحدة">مرجع سريع من شاشة واحدة</h2>
<h4 id="ما-صرنا-نستطيع-فعله">ما صرنا نستطيع فعله</h4>
<p dir="rtl" style="text-align:right">نستطيع تشغيل خلية <code dir="ltr">Julia</code>، وتقييم تعابير قياسية، وتعيين القيم وتغييرها، وطباعة نتيجة موسومة، والتمييز بين أخطاء الصياغة وأخطاء وقت التشغيل والأخطاء المنطقية. ونستطيع قراءة ملف <code dir="ltr">CSV</code> مُعطى إلى <code dir="ltr">DataFrame</code>، واختيار أعمدته المسمّاة، ونسخ وصفة ملاءمة مُعطاة وتفسيرها دون أن ندّعي إتقان المصفوفات أو الملاءمة.</p>
<h4 id="الأفكار-الأساسية-ومرجع-الصياغة">الأفكار الأساسية ومرجع الصياغة</h4>
<p dir="rtl" style="text-align:right">يُنتج <em>التعبير</em> قيمة. و<em>المتغيّر</em> اسمٌ لقيمة. في <code dir="ltr">name = expression</code> تقيّم <code dir="ltr">Julia</code> الطرف الأيمن وتخزّنه تحت الاسم الواقع على اليسار. نستخدم <code dir="ltr">println("label", value)</code> للحصول على مخرجات مقروءة. خطأ الصياغة لا يمكن تحليله؛ وخطأ وقت التشغيل يحدث أثناء التقييم؛ والخطأ المنطقي يعمل لكنه يجيب عن السؤال الخطأ.</p>
<h4 id="الأدوات-التي-نستخدمها-اليوم">الأدوات التي نستخدمها اليوم</h4>
<p dir="rtl" style="text-align:right">القيم القياسية، والمؤثرات <code dir="ltr">+</code> و<code dir="ltr">-</code> و<code dir="ltr">*</code> و<code dir="ltr">/</code>، والتعيين بـ<code dir="ltr">=</code>، و<code dir="ltr">println</code>، و<code dir="ltr">CSV.read</code>، و<code dir="ltr">DataFrame</code>، والوصول إلى عمود بالاسم، وتعليقات الوحدات، وفحص الحالة المعروفة <span dir="ltr" class="math inline">\(V=0\Rightarrow I=0\)</span>. تحتوي وصفة الواجب المُعطاة أيضًا على <code dir="ltr">length</code> و<code dir="ltr">sum</code> والمؤثرات المنقّطة؛ وهذه ليست مهارات مستقلة اليوم.</p>
<h4 id="فحص-قبل-البدء">فحص قبل البدء</h4>
<p dir="rtl" style="text-align:right">دون تشغيل الكود، توقّع القيمة العددية ووحدتها لـ<code dir="ltr">resistance</code> بعد <code dir="ltr">voltage = 2.0</code> و<code dir="ltr">current = 0.4</code> و<code dir="ltr">resistance = voltage / current</code>.</p>
<p dir="rtl" style="text-align:right">للحصول على الشروح الكاملة، راجع الفصل 1 من المحاضرة، ولا سيما قسم <em>أين نرى النتيجة؟</em>، وقسم <em><code dir="ltr">Julia</code> كآلة حاسبة</em>، وقسم <em>أسماء للقيم قبل الفيزياء</em>، وقسم <em>سير عمل موجَّه من القياس إلى النموذج</em>، مع صناديق المربعات الصغرى والبواقي والوحدات وخطأ النموذج.</p>
<h2 id="الاستوديو-الموجه-17-دقيقة">الاستوديو الموجَّه: 17 دقيقة</h2>
<h4 id="صندوق-فيزياء-للطلاب-نابض">صندوق فيزياء للطلاب: نابض</h4>
<p dir="rtl" style="text-align:right">لنابض مُمدَّد مسافة متواضعة داخل نطاقه المرن، يكون مقدار قوة الاستعادة <span dir="ltr" class="math inline">\(F=kx\)</span>. هنا يُقاس التمدد <span dir="ltr" class="math inline">\(x\)</span> من الطول دون حِمل بالمتر، والقوة <span dir="ltr" class="math inline">\(F\)</span> مقدار القوة المؤثّرة بالنيوتن، والميل <span dir="ltr" class="math inline">\(k\)</span> ثابت النابض بوحدة نيوتن لكل متر. نستخدم تمددًا موجبًا ومقدار قوة مؤثّرة موجبًا. عند <span dir="ltr" class="math inline">\(x=0\)</span> يتوقّع النابض المثالي <span dir="ltr" class="math inline">\(F=0\)</span>. وقد يُبطل وصف الخط المستقيم وجود حمل مسبق أو إزاحة أو تخلف هيستيري أو تشوّه لدن أو خروج من النطاق المرن.</p>
<h3 id="دقائق-قراءة-ملف-csv-معطى-والتوقع">0–5 دقائق: قراءة ملف <code dir="ltr">CSV</code> مُعطى والتوقّع</h3>
<p dir="rtl" style="text-align:right">نفتح <code dir="ltr">assets/data/spring_force.csv</code>. ملف <code dir="ltr">CSV</code> جدول نصي تفصل بين قيمه فواصل، وتأتي أسماء الأعمدة في سطره الأول. نقرأه صراحةً إلى <code dir="ltr">DataFrame</code>: جدول تبقى أعمدته المسمّاة متوافقة صفًّا بصف.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

In [2]:
using CSV, DataFrames
spring = CSV.read(joinpath(course_root, "assets", "data", "spring_force.csv"), DataFrame)
x = spring.extension_m
y = spring.force_N
println(length(x), " measurements")

6 measurements

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">مخرجات مرجعية في الملاحظات المكتوبة (للمقارنة مع مخرجات الخلية الفعلية):</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">6 measurements</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">نشغّل مثال قراءة الملف من مجلد بيانات المقرر، <code dir="ltr">assets/data</code>. يحتوي المتجهان <code dir="ltr">x</code> و<code dir="ltr">y</code> على ستّ قياسات متوافقة. نتوقّع <span dir="ltr" class="math inline">\(F=0\)</span> عند <span dir="ltr" class="math inline">\(x=0\)</span> واتجاهًا مستقيمًا متزايدًا. يأتي تنظيف الجداول الأغنى لاحقًا.</p>
<h3 id="دقيقة-ملاءمة-موجهة-واحدة">5–13 دقيقة: ملاءمة موجَّهة واحدة</h3>
<p dir="rtl" style="text-align:right">تستخدم الدفتر المُعطى وصفة المجموع الممركز نفسها التي في المحاضرة. نقرأ نتيجته بدل إعادة تصميمه.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

In [3]:
x = [0.00, 0.02, 0.04, 0.06, 0.08, 0.10]   # m
y = [0.01, 0.39, 0.81, 1.18, 1.61, 1.99]    # N
n = length(x)
xbar = sum(x) / n; ybar = sum(y) / n
slope = sum((x .- xbar) .* (y .- ybar)) / sum((x .- xbar).^2)
intercept = ybar - slope * xbar
println("slope = $(round(slope; digits=1)) N/m")
println("intercept = $(round(intercept; digits=4)) N")

slope = 19.9 N/m


intercept = 0.0033 N


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">مخرجات مرجعية في الملاحظات المكتوبة (للمقارنة مع مخرجات الخلية الفعلية):</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">slope = 19.9 N/m
intercept = 0.0033 N</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">للميل وحدة نيوتن لكل متر، ولذلك يقدّر <span dir="ltr" class="math inline">\(k\approx19.9\ \mathrm{N/m}\)</span>. أما صغر التقاطع فهو فحص للإزاحة، لا قانون جديد.</p>
<p dir="rtl" style="text-align:right">لو غيّرنا <code dir="ltr">voltage</code> إلى <code dir="ltr">3.0</code> دون إعادة تشغيل التعيين الأخير، لبقي <code dir="ltr">resistance</code> على قيمته القديمة. التعيين يسجّل نتيجةً حين يُنفَّذ السطر.</p>
<h3 id="دقيقة-الفحص-والتفسير">13–17 دقيقة: الفحص والتفسير</h3>
<p dir="rtl" style="text-align:right">نتحقّق مما إذا كانت البواقي تقع على جانبي الصفر. ثم نغيّر قيمة قوة مُعطاة واحدة ونعيد الملاءمة: لأن الميل والتقاطع قد يتغيّران، نعيد حساب كل توقّع وكل باقٍ قبل أن نقرّر أي القيم تغيّرت. قد يكون الخط مفيدًا في هذا النطاق دون إثبات أن النابض خطي عند كل تمدد.</p>
<h4 id="نقطة-تحقق">نقطة تحقّق</h4>
<p dir="rtl" style="text-align:right">ما وحدة الميل؟ ولماذا يُعدّ <span dir="ltr" class="math inline">\(F=0\)</span> عند <span dir="ltr" class="math inline">\(x=0\)</span> فحصًا مفيدًا لحالة معروفة؟</p>
<h3 id="دقيقة-الانتقال-إلى-الواجب">17–20 دقيقة: الانتقال إلى الواجب</h3>
<p dir="rtl" style="text-align:right">نفتح ورقة <em>الواجب 1: ملاءمة السقوط الحر</em> المنفصلة والدفتر المُعطى للسقوط الحر. نبدأ بتحديد <code dir="ltr">t_s</code> و<code dir="ltr">y_m</code> و<code dir="ltr">t_s .^ 2</code> والتوقّعات والبواقي.</p>
<h4 id="بعد-الدقيقة-17">بعد الدقيقة 17</h4>
<p dir="rtl" style="text-align:right">انتهى الجزء الموجَّه من التمرين. ويُخصَّص ما تبقّى لعمل مدعوم على الواجب الأساسي المنفصل وحده. لا نضيف تمرينًا آخر، ولا نفتح المكافأة قبل إتمام الأساس باستقلال.</p>
<h2 id="التمرين-قياس-ثابت-النابض">التمرين: قياس ثابت النابض</h2>
<p dir="rtl" style="text-align:right">ننقل التحليل نفسه إلى نابض. قياساتنا هي الاستطالة <span dir="ltr" class="math inline">\(x\)</span> بالمتر والقوة المؤثرة <span dir="ltr" class="math inline">\(F\)</span> بالنيوتن؛ والنموذج الملائم هو <span dir="ltr" class="math inline">\(F=kx+F_0\)</span>. وخلافًا لملاءمة تكون فيها الاستطالة على المحور الرأسي، يكون الميل هنا هو ثابت النابض <span dir="ltr" class="math inline">\(k\)</span> بوحدة N/m مباشرة. ونتوقّع أولًا ميلًا موجبًا ونتحقق من أن الحد الثابت له وحدة القوة.</p>
<p dir="rtl" style="text-align:right">قبل الجلسة نفتح دفتر تمرين الأسبوع الأول ونحدّد خلاياه المُعطاة وخلايا الملاءمة. وفي أول عشرين دقيقة نفحص المحاور، ونتابع ملاءمة واحدة، ونشرح الميل والحد الثابت وباقيًا واحدًا. ثم نبدأ واجب السقوط الحر المستقل بمساعدة طاقم التدريس. ومن الفحوص المفيدة أن نشرح لماذا يغيّر تبديل المحورين وحدات الميل الملائم ومعناه؛ وفي البيانات المشوّشة، لا تكافئ إعادة الملاءمة بمحورين مبدَّلين عمومًا مقلوب الميل.</p>
<h2 id="الواجب-الأساسي-استنتاج-الجاذبية-من-قياسات-السقوط-الحر">الواجب الأساسي: استنتاج الجاذبية من قياسات السقوط الحر</h2>
<p dir="rtl" style="text-align:right">نستقصي هل تتفق مسافات السقوط المقيسة مع الإطلاق من السكون تحت جاذبية منتظمة تقريبًا. نأخذ المسافة موجبة نحو الأسفل، فيكون <span dir="ltr" class="math inline">\(y(t)=\tfrac12gt^2+b\)</span>، مع <span dir="ltr" class="math inline">\(g&gt;0\)</span> بوحدة m/s<span dir="ltr" class="math inline">\(^2\)</span> والحد الثابت <span dir="ltr" class="math inline">\(b\)</span> بالمتر. والتحويل المُعطى <code dir="ltr">t_s .^ 2</code> يجعل الزمن المربّع هو الإحداثي الأفقي. نلائم <span dir="ltr" class="math inline">\(y=a t^2+b\)</span> ونفسّر <span dir="ltr" class="math inline">\(g=2a\)</span>؛ وعامل الاثنين يأتي من معادلة الحركة، لا من اصطلاح برمجي.</p>
<p dir="rtl" style="text-align:right">نعمل على هذه المراحل:</p>
<ol type="1">
<li dir="rtl" style="text-align:right"><p dir="rtl" style="text-align:right">نفحص الملف المُعطى <code dir="ltr">free_fall.csv</code>: نحدّد عمودَي الزمن والمسافة ووحدتيهما وعدد المشاهدات.</p></li>
<li dir="rtl" style="text-align:right"><p dir="rtl" style="text-align:right">نتوقّع شكل <span dir="ltr" class="math inline">\(y\)</span> بدلالة <span dir="ltr" class="math inline">\(t\)</span> وشكل <span dir="ltr" class="math inline">\(y\)</span> بدلالة <span dir="ltr" class="math inline">\(t^2\)</span> قبل تشغيل الملاءمة. نطبّق الوصفة المُعطاة ونحتفظ بالتنبؤات وبواقي المقيس ناقص المتوقّع.</p></li>
<li dir="rtl" style="text-align:right"><p dir="rtl" style="text-align:right">ننشئ شكلًا واحدًا: النقاط المقيسة كعلامات مع النموذج الملائم فوقها، وكل محور مسمّى بكميته ووحدته.</p></li>
<li dir="rtl" style="text-align:right"><p dir="rtl" style="text-align:right">نحفظ الشكل بصيغة <code dir="ltr">PNG</code> أو <code dir="ltr">PDF</code> ونذكر <span dir="ltr" class="math inline">\(g\)</span> والحد الثابت الملائمين. ونشرح هل تدعم البواقي النموذج على المجال المقيس، ونسمّي افتراضًا واحدًا لا تستطيع الملاءمة إثباته.</p></li>
<li dir="rtl" style="text-align:right"><p dir="rtl" style="text-align:right">نقدّم السكربت والشكل على صفحة المقرر في <code dir="ltr">Moodle</code>، كما هو محدَّد في ورقة الواجب.</p></li>
</ol>
<p dir="rtl" style="text-align:right">نقدّم سكربتًا كاملًا مع الشكل؛ وتعريف دالة <code dir="ltr">Julia</code> ليس جزءًا من عمل الأسبوع الأول. ويُصحَّح الواجب الأول يدويًا في <code dir="ltr">Moodle</code>: ناجح/راسب. ومن المفيد أن نُبقي ورقة الواجب بجانبنا، فهي المكان الذي تُذكر فيه الأعمدة المطلوبة والشكل والكميات التي يجب الإبلاغ عنها. ويجب ألّا يؤدي تقريب إجابة معروضة إلى تقريب القيم المستخدمة في الحساب.</p>
<h2 id="مكافأة-اختيارية-اجعل-رسما-علميا-يوصل-الفكرة">مكافأة اختيارية: اجعل رسمًا علميًا يوصّل الفكرة</h2>
<p dir="rtl" style="text-align:right">هذه المكافأة اختيارية، وليست شرطًا مسبقًا لعمل لاحق، وغير ممتحنة. وكل مكافأة نُنجزها تساوي <strong>ثلث نقطة من العلامة النهائية</strong>.</p>
<p dir="rtl" style="text-align:right">بعد قراءة مستقلة مدتها 10–15 دقيقة، نحسّن رسمًا أوهليًا مصغّرًا مُعطى. هنا الجهد هو الإحداثي الأفقي والتيار هو الإحداثي الرأسي، لذا يكون ميل النموذج هو الموصلية بوحدة S لا المقاومة بوحدة <span dir="ltr" class="math inline">\(\Omega\)</span>. ونحدّد سلسلة التيار المقيس وسلسلة التيار الملائم قبل تغيير مظهرهما.</p>
<p dir="rtl" style="text-align:right">نضيف العنوان المطلوب، والتسميات مع الوحدات، ومدخلات مفتاح الرسم، وعلامات المشاهدات، وخط النموذج، والإطار، وحجم اللوحة، ثم نصدّر <code dir="ltr">PNG</code> و<code dir="ltr">PDF</code>. وتوفّر الشيفرة العددية السلسلتين المتوافقتين المتحقَّق منهما؛ وتُقيَّم الصور منفصلة وفق قائمة تحقّق صريحة. ونتأكد أن تغيير تسمية أو نمط خط لا يغيّر قياسًا. وتحدّد ورقة المكافأة السلاسل النصية وأسماء الملفات بدقة، فيكون الاكتمال، لا الذوق ولا مطابقة البكسل، هو معيار النجاح.</p>
<h2 id="استكشاف-إضافي-منحنيات-ليساجو">استكشاف إضافي: منحنيات ليساجو</h2>
<p dir="rtl" style="text-align:right">في هذا الاستقصاء الاختياري غير الممتحن نستخدم وصفة رسم مُعطاة لـ <span dir="ltr" class="math display">\[x(t)=A\cos(at+\delta),\qquad y(t)=B\sin(bt).\]</span> هنا <span dir="ltr" class="math inline">\(t\)</span> بالثواني، و<span dir="ltr" class="math inline">\(a,b\)</span> ترددان زاويان بوحدة rad/s، و<span dir="ltr" class="math inline">\(\delta\)</span> طور بالراديان، و<span dir="ltr" class="math inline">\(A,B\)</span> يحدّدان سعتي الإحداثيين. وتصف هذه المنحنيات حركتين جيبيتين مستقلتين ننظر إليهما معًا؛ وهي ليست مسارات تحت قوة جاذبية مركزية.</p>
<p dir="rtl" style="text-align:right">نُبقي السعتين ثابتتين ونقارن نسخ الترددات <span dir="ltr" class="math inline">\(1:1\)</span> و<span dir="ltr" class="math inline">\(2:1\)</span> و<span dir="ltr" class="math inline">\(3:2\)</span>. ثم، لنسخة واحدة، نغيّر الطور من <span dir="ltr" class="math inline">\(0\)</span> إلى <span dir="ltr" class="math inline">\(\pi/2\)</span>. نحفظ رسمين متباينين ونكتب تخمينًا عن الإغلاق والتناظر. ونتحقق هل يغيّر مدى زمني أطول منحنى يبدو مفتوحًا: فنافذة الرسم المحدودة ليست دليلًا على عدم الإغلاق. ولا يضيف هذا النشاط أي صياغة برمجية مطلوبة، وهو منفصل عن المكافأة المُقيَّمة.</p>
</div>

<div dir="rtl" style="text-align:right">
<h2 id="أمثلة-صغيرة-وفحوص-مذكورة-في-الفصل">أمثلة صغيرة وفحوص مذكورة في الفصل</h2>
<p dir="rtl" style="text-align:right">تجعل الخلايا التالية الأمثلة المضمّنة والفحوص المحلولة في الفصل قابلة للتنفيذ. وهي لا تضيف حلولًا للواجب.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">فحص تمهيدي: قسمة الجهد على التيار تعطي وحدات المقاومة. تغيير الجهد لا يعيد حساب مقاومة مخزَّنة.</p>
</div>

In [4]:
voltage = 2.0      # V
current = 0.4      # A
resistance = voltage / current   # ohm
println(resistance)
voltage = 3.0      # V
println(resistance)

5.0


5.0


<div dir="rtl" style="text-align:right">
<h2 id="الختام-والتأمل">الختام والتأمل</h2>
<p dir="rtl" style="text-align:right">نشرح نتيجة واحدة بالكلمات، ونذكر وحداتها أو افتراضاتها حين يلزم، ونحدّد فحصًا قد يكشف خطأ. تصف فقرات التمرين والواجب والمكافأة الاختيارية والاستزادة أعلاه الخطوات التالية. لا تصبح الخلفية الاختيارية شرطًا للامتحان.</p>
</div>